core/domain/caching_services.py

```diff
@@ -0,0 +1,263 @@
+# object). This namespace does not support sub-namespaces.
+CACHE_NAMESPACE_CONFIG = 'config'
+# The sub-namespace is not necessary for the default namespace. The namespace
+# handles default datatypes allowed by Redis including Strings, Lists, Sets,
+# and Hashes. More details can be found at: https://redis.io/topics/data-types.
+CACHE_NAMESPACE_DEFAULT = 'default'
+
+DESERIALIZATION_FUNCTIONS = {
+    CACHE_NAMESPACE_COLLECTION: collection_domain.Collection.deserialize,
+    CACHE_NAMESPACE_EXPLORATION: exp_domain.Exploration.deserialize,
+    CACHE_NAMESPACE_SKILL: skill_domain.Skill.deserialize,
+    CACHE_NAMESPACE_STORY: story_domain.Story.deserialize,
+    CACHE_NAMESPACE_TOPIC: topic_domain.Topic.deserialize,
+    CACHE_NAMESPACE_PLATFORM: (
+        platform_parameter_domain.PlatformParameter.deserialize),
+    CACHE_NAMESPACE_CONFIG: lambda x: x,
+    CACHE_NAMESPACE_DEFAULT: lambda x: x
+}
+
+SERIALIZATION_FUNCTIONS = {
+    CACHE_NAMESPACE_COLLECTION: lambda x: x.serialize(),
+    CACHE_NAMESPACE_EXPLORATION: lambda x: x.serialize(),
+    CACHE_NAMESPACE_SKILL: lambda x: x.serialize(),
+    CACHE_NAMESPACE_STORY: lambda x: x.serialize(),
+    CACHE_NAMESPACE_TOPIC: lambda x: x.serialize(),
+    CACHE_NAMESPACE_PLATFORM: lambda x: x.serialize(),
+    CACHE_NAMESPACE_CONFIG: lambda x: x,
+    CACHE_NAMESPACE_DEFAULT: lambda x: x
+}
+
+
+def _get_memcache_key(namespace, sub_namespace, obj_id):
+    """Returns a memcache key for the class under the corresponding
+    namespace and sub_namespace.
+
+    Args:
+        namespace: str. The namespace under which the values associated with the
+            id lie. Use CACHE_NAMESPACE_DEFAULT as the namespace for ids that
+            are not associated with a conceptual domain-layer entity and
+            therefore don't require serialization.
+        sub_namespace: str|None. The sub-namespace further differentiates the
+            values. For Explorations, Skills, Stories, Topics, and Collections,
+            the sub-namespace is the stringified version number of the objects.
+        obj_id: str. The id of the value to store in the memory cache.
+
+    Raises:
+        Exception. The sub-namespace contains a ':'.
+
+    Returns:
+        str. The generated key for use in the memory cache in order to
+        differentiate a passed-in key based on namespace and sub-namespace.
+    """
+    sub_namespace_key_string = (sub_namespace or '')
+    if MEMCACHE_KEY_DELIMITER in sub_namespace_key_string:
+        raise ValueError(
+            'Sub-namespace %s cannot contain \':\'.' % sub_namespace_key_string)
+    return '%s%s%s%s%s' % (
+        namespace, MEMCACHE_KEY_DELIMITER,
+        sub_namespace_key_string, MEMCACHE_KEY_DELIMITER, obj_id)
+
+
+def flush_memory_cache():
+    """Flushes the memory cache by wiping all of the data."""
+    memory_cache_services.flush_cache()
+
+
+def get_multi(namespace, sub_namespace, obj_ids):
```

Comment:
1. No, I don't think so. The suggested changes encourage the use of caching_services a general-purpose thing for caching anything -- and this is actually not something we want to do, because it's hard to get caching right. So the changes made here are with the aim of hardening the API.
- Not all sub_namespaces are version numbers.
- Prefer not having default values, in general; they lead to branching in the code. In particular, with something like memcache, it is important for callers to know exactly what they are doing -- in the past, I've seen subtle bugs that are hard to track down when the use of the cache is not well-thought-through.
- See general point above: the aim is not to make it general-purpose. We want this to be used only for specific things.

@DubeySandeep -- what do you think?

2. I don't understand your question, sorry. Could you elaborate, please? The obj_id should disambiguate which item is to be fetched.

Reply:
> @DubeySandeep -- what do you think?

I'm fine with what you have suggested. Thanks!

Your answer also explains the 2nd bullet.